torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -214,23 +214,19 @@ void ReplayTransformations::runReplay() {
       [](std::pair<IterDomain*, size_t> entry) { return entry.first; });
 }
 
-// TODO: Make sure the replay and target domains have a
-// producer-consumer relationship when forward_bcast_mismatch is
-// true. When it's true, a merge expr with amissing axis may
-// erroneously be forwarded even if the axis of the replayed tensor is
-// not broadcast. It should not occur when the replay and target
-// domains have a producer-consumer relationship.
 BestEffortReplay::BestEffortReplay(
     const std::vector<IterDomain*>& replay_domain,
     const std::vector<IterDomain*>& target_domain,
-    std::unordered_map<IterDomain*, IterDomain*> replay_map,
-    bool forward_bcast_mismatch)
-    : id_map_(std::move(replay_map)) {
-  for (auto entry : id_map_)
+    std::unordered_map<IterDomain*, IterDomain*> target2replay_map,
+    std::unordered_map<IterDomain*, IterDomain*> forward_id_map)
+    : target2replay_id_map_(std::move(target2replay_map)),
+      forward_id_map_(std::move(forward_id_map)) {
```

Comment:
Wouldn't it make sense to have a C++ test to exercise this new way of forwarding logic? I think it would be great to have some test that doesn't work with the current naive approach but works with this PR.

In any case, it's really great to have this more robust forwarding logic!

Reply:
I was going to, but norm_hack_2 schedulers exercise this path pretty well so I thought switching to those schedulers would suffice for now. I don't catch all corner cases, but I'm not sure we really need to generally, so intended to be a bit more passive on that front. Much of this was for BatchNorm so when we get better support of that I think testing will be pretty thorough.

Reply:
There was a question on forwarding consumer id's, will add a test exercising that.

Reply:
That would be great. Would be very helpful to understand the changes.

Reply:
Test added: FusionAdvancedComputeAt7_CUDA is the important one.

Reply:
I tried the test with both this PR and the current devel. The generated kernels were identical, except for the use of `barrier.sync`. Would it be difficult to have something that's enabled by this PR? I think that would be a good illustrative example for the new forwarding logic.